<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Genome &mdash; topo_package 1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Transcriptome" href="Transcriptome.html" />
    <link rel="prev" title="Presentation" href="Presentation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            topo_package
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Presentation.html">Presentation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Genome</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#class-methods">class methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-useful_functions_genome">useful functions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Transcriptome.html">Transcriptome</a></li>
<li class="toctree-l1"><a class="reference internal" href="Chipseq.html">Chipseq</a></li>
<li class="toctree-l1"><a class="reference internal" href="HiC.html">HiC</a></li>
<li class="toctree-l1"><a class="reference internal" href="plot_genome.html">plot_genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="GO.html">GO</a></li>
<li class="toctree-l1"><a class="reference internal" href="stat_analysis.html">statistical analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="TSS_TTS_TU.html">TSS_TTS_TU</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">topo_package</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Genome</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/Genome.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="genome">
<h1>Genome<a class="headerlink" href="#genome" title="Permalink to this heading"></a></h1>
<section id="class-methods">
<h2>class methods<a class="headerlink" href="#class-methods" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="Genome.Genome">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Genome.</span></span><span class="sig-name descname"><span class="pre">Genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Genome.Genome" title="Permalink to this definition"></a></dt>
<dd><p>The Genome class is one of the main classes of this package. It gathers all
the attributes of a genome such as the sequence, the set of genes (with their
functional annotations and their orientation) but also all the annotations of
TSS, TU and TTS.</p>
<p>Each Genome instance has to be initialized with an organism name</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Genome.Genome.load_seq">
<span class="sig-name descname"><span class="pre">load_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sequence.fasta'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Genome.Genome.load_seq" title="Permalink to this definition"></a></dt>
<dd><p>Load_seq loads DNA sequence from a .fasta file present in the
main directory of the organism using useful_functions_genome.load_seq
function. Adds this sequence, its complement, and its length to a
Genome instance.
Creates 3 new attributes to the Genome instance</p>
<blockquote>
<div><ul class="simple">
<li><p>seq (str.): genomic sequence compose of A,T,G and C</p></li>
<li><p>seqcompl (str.): complement sequence to seq</p></li>
<li><p>length (int.): length of the genomic sequence</p></li>
</ul>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>filename</strong> (<em>Optional</em><em> [</em><em>str.}</em>) – name of the file containing the DNA
sequence in FASTA format. Default: “sequence.fasta”</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_seq</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="mi">100</span><span class="p">:</span><span class="mi">151</span><span class="p">]</span>
<span class="go">&#39;AATGTCGATCTTCAACATATCGCCGATCCGACGGGCACCCAGATCCTGCAG&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">seqcompl</span><span class="p">[</span><span class="mi">100</span><span class="p">:</span><span class="mi">151</span><span class="p">]</span>
<span class="go">&#39;TTACAGCTAGAAGTTGTATAGCGGCTAGGCTGCCCGTGGGTCTAGGACGTC&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Genome.Genome.load_annotation">
<span class="sig-name descname"><span class="pre">load_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annot_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sequence.gff3'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Genome.Genome.load_annotation" title="Permalink to this definition"></a></dt>
<dd><p>load_annotation loads a gene annotation (coordinates, length,
name…) from a file present in the /annotation/ directory.</p>
<dl class="simple">
<dt>Creates a new attribute of the Genome instance:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.genes (dict.)</dt><dd><p>self.genes is a dictionary of shape
{locus_tags: Gene object}. Each Gene object is
initialized with the following attributes:
locus_tag, ID, name,strand, left, right, start, end,
middle, length and ASAP.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>filename</strong> (<em>Optional</em><em> [</em><em>str.</em><em>]</em>) – name of the file containing the
genomic annotation. Default: “sequence.gff3”</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If the file is a .gff3 or .gff information will be loaded using
useful_functions_genome.load_gff.
Else, the information importation requires an annotation.info file,
containing column indices of each information in the data file and
some additional information, in the following order:
[0] Filename [1] Separator [2] Locus_tag column [3] Name column
[4] ID column [5] Strand column [6] Left coordinate column
[7] Right coordinate column [8] File start line</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_annotation</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="s2">&quot;Dda3937_00005&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">name</span>
<span class="go">&#39;guaA&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Genome.Genome.load_genes_per_pos">
<span class="sig-name descname"><span class="pre">load_genes_per_pos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Genome.Genome.load_genes_per_pos" title="Permalink to this definition"></a></dt>
<dd><p>load_genes_per_pos associates each position with a list of genes
overlapping this and its surrounding positions delimited
by the window size given as an argument.</p>
<dl class="simple">
<dt>Creates a new attribute of the Genome instance:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.genes_per_pos (dict.)</dt><dd><p>self.genes is a dictionary of shape {position: list of
genes}. It contains, for each position p, the list of
genes overlapping any position between p-window/2 and
p+window/2 (inclusive)</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>window</strong> (<em>Optional</em><em> [</em><em>int.</em><em>]</em>) – window size in b. load_genes_per_pos
finds genes between pos-window/2 and pos+window/2
(inclusive). Default: window = 0</p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic annotation. If no annotation is
loaded, the load_annotation method with the default “sequence.gff3”
file is computed. To use another annotation, please load an
annotation before using the load_genes_per_pos method.</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_annotation</span><span class="p">(</span><span class="n">annot_file</span><span class="o">=</span><span class="s2">&quot;sequence.gff3&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_genes_per_pos</span><span class="p">(</span><span class="n">window</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">genes_per_pos</span><span class="p">[</span><span class="mi">120</span><span class="p">]</span>
<span class="go">[&#39;Dda3937_00155&#39;, &#39;Dda3937_00156&#39;, &#39;Dda3937_00154&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Genome.Genome.load_neighbor_all">
<span class="sig-name descname"><span class="pre">load_neighbor_all</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Genome.Genome.load_neighbor_all" title="Permalink to this definition"></a></dt>
<dd><p>For each gene and positions, load_neighbor_all finds nearest
neighbors (left and right) on genome, whatever their strand.</p>
<dl class="simple">
<dt>Creates 2 new attributes of Gene instances:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.genes[locus].left_neighbor</dt><dd><p>locus of the nearest left-side neighbor gene</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.genes[locus].right_neighbor</dt><dd><p>locus of the nearest right-side neighbor gene</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>and 4 new attributes of Genome instance:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.genomic_situation (dict.)</dt><dd><p>dict of shape {position: situation} with situation either
“intergenic” or “intragenic”.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.left_neighbor (dict.)</dt><dd><p>dict of shape {position: locus of the nearest left-side
neighbor gene}</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.right_neighbor (dict.)</dt><dd><p>dict of shape {position: locus of the nearest right-side
neighbor gene}</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.gene (dict.)</dt><dd><p>dict of shape {position: gene} with gene = “NA” if the
position is intergenic.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic annotation. If no annotation is
loaded, the load_annotation method with the default “sequence.gff3”
file is computed. To use another annotation, please load an
annotation before using the load_neighbor_all method.</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_annotation</span><span class="p">(</span><span class="n">annot_file</span><span class="o">=</span><span class="s2">&quot;sequence.gff3&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_neighbor_all</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="s2">&quot;Dda3937_00005&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">left_neighbor</span>
<span class="go">&#39;Dda3937_00006&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="s2">&quot;Dda3937_00005&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">right_neighbor</span>
<span class="go">&#39;Dda3937_00004&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">g</span><span class="o">.</span><span class="n">genomic_situation</span><span class="p">[</span><span class="mi">569</span><span class="p">]</span>
<span class="go">&#39;intragenic&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">gene</span><span class="p">[</span><span class="mi">569</span><span class="p">]</span>
<span class="go">&#39;Dda3937_00156&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">left_neighbor</span><span class="p">[</span><span class="mi">569</span><span class="p">]</span>
<span class="go">&#39;Dda3937_00155&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">right_neighbor</span><span class="p">[</span><span class="mi">569</span><span class="p">]</span>
<span class="go">&#39;Dda3937_00157&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Genome.Genome.load_gene_orientation">
<span class="sig-name descname"><span class="pre">load_gene_orientation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">couple</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Genome.Genome.load_gene_orientation" title="Permalink to this definition"></a></dt>
<dd><p>Compute gene orientation with the following criteria:</p>
<ul>
<li><p>If couple = 3, gene is considered:</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>divergent</cite> if left neighbor on - strand and right neighbor on + strand,</p></li>
<li><p><cite>convergent</cite> if left neighbor on + strand and right neighbor on - strand,</p></li>
<li><p><cite>tandem</cite> if left and right neighbors on same strand (whatever the strand of the given gene is),</p></li>
<li><p><cite>isolated</cite> if the distance between neighbors is higher than the maximal distance given as argument.</p></li>
</ul>
</div></blockquote>
</li>
<li><p>If couple = 2, gene is considered:</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>tandem</cite> if predecessor (left neighbor for gene on + strand, right neighbor for gene on - strand) is on same strand,</p></li>
<li><p><cite>divergent</cite> if the predecessor is on opposite strand.</p></li>
</ul>
</div></blockquote>
</li>
</ul>
<dl class="simple">
<dt>Creates new attributes:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.orientation (dict.)</dt><dd><p>new attribute of the Genome instance.
self.orientation is a dictionary of shape {orientation:
list of genes}. It contains the list of genes for each
orientation (tandem, divergent, convergent, and isolated
if couple=3, tandem and divergent if couple=2)</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.genes[locus].orientation (str.)</dt><dd><p>new attribute of Gene instances related to the Genome instance
given as argument</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>couple</strong> (<em>int.</em>) – <p>number of genes to consider in a “couple”.</p>
<ul>
<li><p>If couple = 2: computes the orientation of a gene
relative to its predecessor</p></li>
<li><p>If couple = 3: computes the orientation of a gene
relative to its two neighbors</p></li>
</ul>
<p>Default: 3</p>
</p></li>
<li><p><strong>max_dist</strong> (<em>Optional</em><em> [</em><em>int.</em><em>]</em>) – maximal distance between 2 genes
start positions for seeking neighbor (Default: 5kb)</p></li>
</ul>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic annotation. If no annotation is
loaded, the load_annotation method with the default “sequence.gff3”
file is computed. To use another annotation, please load an
annotation before using the load_neighbor_all method.</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_annotation</span><span class="p">(</span><span class="n">annot_file</span><span class="o">=</span><span class="s2">&quot;sequence.gff3&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_gene_orientation</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="s2">&quot;Dda3937_00005&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">orientation</span>
<span class="go">&#39;tandem&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">orientation</span><span class="p">[</span><span class="s2">&quot;isolated&quot;</span><span class="p">]</span>
<span class="go">[&#39;Dda3937_02360&#39;, &#39;Dda3937_01107&#39;, &#39;Dda3937_03898&#39;, &#39;Dda3937_04216&#39;,</span>
<span class="go">&#39;Dda3937_00244&#39;, &#39;Dda3937_04441&#39;, &#39;Dda3937_01704&#39;, &#39;Dda3937_01705&#39;,</span>
<span class="gp">...</span>
<span class="go">&#39;Dda3937_02126&#39;, &#39;Dda3937_01530&#39;, &#39;Dda3937_04419&#39;, &#39;Dda3937_02081&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Genome.Genome.load_pos_orientation">
<span class="sig-name descname"><span class="pre">load_pos_orientation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">max_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Genome.Genome.load_pos_orientation" title="Permalink to this definition"></a></dt>
<dd><p>Computes gene orientation with the following criteria:</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>divergent</cite> if left neighbor on - strand and right neighbor on + strand,</p></li>
<li><p><cite>convergent</cite> if left neighbor on + strand and right neighbor on - strand,</p></li>
<li><p><cite>tandem</cite> if left and right neighbors on same strand</p></li>
<li><p><cite>isolated</cite> if the distance between neighbors is higher than the maximal distance given as argument</p></li>
</ul>
</div></blockquote>
<dl class="simple">
<dt>Creates 2 new attributes of the Genome instance:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.pos_orientation (dict of dict)</dt><dd><p>dictionary containing 2 subdictionaries.
One subdictionary for “intergenic” positions and one for
“intragenic” positions.
Each subdictionary contains the list of position for
each orientation. {“intergenic”:{orientation: list of
positions}}, “intragenic”:{orientation: list of positions}}
with orientation in [“divergent”,”convergent”,”tandem”,
“isolated”]</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.orientation_per_pos (dict.)</dt><dd><p>dictionary of shape {position: orientation}.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>max_dist</strong> (<em>Optional</em><em> [</em><em>int.</em><em>]</em>) – maximal distance between 2 genes
start positions for seeking neighbor (Default: 5kb)</p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic annotation. If no annotation is
loaded, the load_annotation method with the default “sequence.gff3”
file is computed. To use another annotation, please load an
annotation before using this method.</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_annotation</span><span class="p">(</span><span class="n">annot_file</span><span class="o">=</span><span class="s2">&quot;sequence.gff3&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_pos_orientation</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">orientation_per_pos</span><span class="p">[</span><span class="mi">30</span><span class="p">]</span>
<span class="go">&#39;tandem&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">pos_orientation</span><span class="p">[</span><span class="s2">&quot;intragenic&quot;</span><span class="p">][</span><span class="s2">&quot;isolated&quot;</span><span class="p">]</span>
<span class="go">[11534,11535,11536,11537,11538,11539,11540,11541,11542,11543,...]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Genome.Genome.load_TSS">
<span class="sig-name descname"><span class="pre">load_TSS</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Genome.Genome.load_TSS" title="Permalink to this definition"></a></dt>
<dd><p>load_TSS loads a TSS annotation from a file present in the /TSS/
directory.</p>
<dl class="simple">
<dt>Creates:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.TSSs (dict. of dict.)</dt><dd><p>New attribute of the Genome instance
self.TSSs is a dictionary of shape
{Condition: {TSS position: TSS object}}.
One subdictionary is created for each condition listed in
TSS.info file. Each TSS object is initialized with the
following attributes: pos, genes, promoter, score, strand
The promoter attribute is a dictionary containing, for
each sigma factor (keys) a subdictionary (value). The
first created key of this subdictionary is “sites” and
the associated value is a tuple containing the positions
of promoter elements. See __init__ and add_promoter in
the TSS class for more details about each attribute.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.TSSs[‘all_TSSs’] (subdictionary of self.TSSs)</dt><dd><p>additional subdictionary of self.TSSs, of shape:
self.TSSs[‘all_TSSs’]={TSSpos: [TSScond]} with [TSScond]
the list of TSS conditions where this TSS was found.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The information importation requires a TSS.info file,
containing column indices of each information in the data file and
some additional information, in the following order:
[0] Condition [1] Filename [2] Locus tag column [3] TSS position
[4] File start line [5] Separator [6] Strand column
[7] Sigma factor column [8] Sites column [9] Score column</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If some of the data types are missing (locus tags, sigma factors,
scores or sites), an empty space can be left in the .info file.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic annotation. If no annotation is
loaded, the load_annotation method with the default “sequence.gff3”
file is computed. To use another annotation, please load an
annotation before using this method.</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_TSS</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">TSSs</span><span class="p">[</span><span class="s2">&quot;dickeya-btss&quot;</span><span class="p">][</span><span class="mi">4707030</span><span class="p">]</span><span class="o">.</span><span class="n">promoter</span>
<span class="go">{&#39;sigma70&#39;: {&#39;sites&#39;: (4707039, 4707044, 4707060, 4707065)},</span>
<span class="go"> &#39;sigma32&#39;: {&#39;sites&#39;: (4707037, 4707046, 4707062, 4707068)}}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">TSSs</span><span class="p">[</span><span class="s2">&quot;dickeya-btss&quot;</span><span class="p">][</span><span class="mi">4707030</span><span class="p">]</span><span class="o">.</span><span class="n">strand</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Genome.Genome.load_prom_elements">
<span class="sig-name descname"><span class="pre">load_prom_elements</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prom_region</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[0,</span> <span class="pre">0]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Genome.Genome.load_prom_elements" title="Permalink to this definition"></a></dt>
<dd><p>load_prom_elements extracts sequences of the different promoter
elements (spacer, -10, -35, discriminator, region around TSS) based
on -35 and -10 coordinates loaded with load_TSS method, for all TSS
conditions. For each sigma factor associated to a TSS annotation, creates
a subdictionary in self.TSSs[condTSS][TSS].promoter with the shape
promoter[sigma] = {element: sequence of the element} with element
in [“spacer”, “minus10”, “minus35”, “discriminator”, “region”].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>shift</strong> (<em>Optional</em><em> [</em><em>int.</em><em>]</em>) – number of nucleotides to include beyond
each region on either side (Default: 0nt)</p></li>
<li><p><strong>prom_region</strong> (<em>Optional</em><em> [</em><em>int.</em><em>,</em><em>int.</em><em>]</em>) – region upstream and
downstream TSSs to extract. Argument with the shape:
[length before TSS, length after TSS].
Default: [0,0] ie no sequence will be extracted around
TSS.</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>See load_TSS description to understand the structure of the
subdictionary self.TSSs[condTSS][TSS].promoter</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic annotation. If no annotation is
loaded, the load_annotation method with the default “sequence.gff3”
file is computed. To use another annotation, please load an
annotation before using this method.</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_prom_elements</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">TSSs</span><span class="p">[</span><span class="s2">&quot;dickeya-btss&quot;</span><span class="p">][</span><span class="mi">4707030</span><span class="p">]</span><span class="o">.</span><span class="n">promoter</span>
<span class="go">{&#39;sigma70&#39;: {&#39;sites&#39;: (4707039, 4707044, 4707060, 4707065),</span>
<span class="go">             &#39;spacer&#39;: &#39;TCGCCCACCCTCAAT&#39;,</span>
<span class="go">             &#39;minus10&#39;: &#39;CATCAT&#39;,</span>
<span class="go">             &#39;minus35&#39;: &#39;TACCCC&#39;,</span>
<span class="go">             &#39;discriminator&#39;: &#39;GAATAACC&#39;},</span>
<span class="go"> &#39;sigma32&#39;: {&#39;sites&#39;: (4707037, 4707046, 4707062, 4707068),</span>
<span class="go">             &#39;spacer&#39;: &#39;CCTCGCCCACCCTCA&#39;,</span>
<span class="go">             &#39;minus10&#39;: &#39;ATCATCATGA&#39;,</span>
<span class="go">             &#39;minus35&#39;: &#39;CCGTACC&#39;,</span>
<span class="go">             &#39;discriminator&#39;: &#39;ATAACC&#39;}}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Genome.Genome.load_TU">
<span class="sig-name descname"><span class="pre">load_TU</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Genome.Genome.load_TU" title="Permalink to this definition"></a></dt>
<dd><p>load_TU loads a TU annotation from a file present in the /TU/
directory.</p>
<dl class="simple">
<dt>Creates a new attribute of the Genome instance:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.TUs (dict. of dict.)</dt><dd><p>self.TUs is a dictionary of shape
{Condition: {TU start pos: TU object}}
One subdictionary is created for each condition listed
in TU.info file. Each TU object is initialized with the
following attributes:  start, stop, orientation, genes,
left,right. See __init__ in the TU class for more details
about each attribute.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The information importation requires a TU.info file,
containing column indices of each information in the data file and
some additional information, in the following order:
[0] Condition [1] Filename [2] Start column [3] Stop column
[4] Strand column [5] Gene column [6] File start line [7] Separator
See the load_TU_cond function in useful_functions_genome for more
details.</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_TU</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">TUs</span><span class="p">[</span><span class="s2">&quot;TU_Forquet&quot;</span><span class="p">][</span><span class="mi">2336702</span><span class="p">]</span><span class="o">.</span><span class="n">right</span>
<span class="go">2339864</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">TUs</span><span class="p">[</span><span class="s2">&quot;TU_Forquet&quot;</span><span class="p">][</span><span class="mi">2336702</span><span class="p">]</span><span class="o">.</span><span class="n">strand</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Genome.Genome.load_TTS">
<span class="sig-name descname"><span class="pre">load_TTS</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Genome.Genome.load_TTS" title="Permalink to this definition"></a></dt>
<dd><p>load_TTS loads a TTS annotation from a file present in the /TTS/
directory.</p>
<dl class="simple">
<dt>Creates:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.TTSs (dict. of dict.)</dt><dd><p>New attribute of the Genome instance
self.TTSs is a dictionary of shape
{Condition: {TTS position: TTS object}}.
One subdictionary is created for each condition listed
in TTS.info file. Each TTS object is initialized with the
following attributes: left, right,  start, end, strand,
rho_dpdt, genes, seq, score.
If the data do not contain information about associated
genes, sequence, or rho dependency, the corresponding
attributes will be initialized as “None”.
See __init__ in the TTS class for more details about
each attribute.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.TTSs[‘all_TTSs’] (subdictionary of self.TTSs)</dt><dd><p>additional subdictionary of self.TTSs, of shape:
self.TTSs[‘all_TTSs’]={TTSpos: [TTScond]}
with [TTScond] the list of TTS conditions where this TTS
was found.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The information importation requires a TTS.info file,
containing column indices of each information in the data file and
some additional information, in the following order:
[0] Condition [1] Filename [2] Left coordinate column
[3] Right coordinate column [4] Strand column [5] Startline
[6] Separator [7] Sequence column
and optionally:
[8] Score column [9] Genes column [10] Rho dependency column</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>See the load_TTS_cond function in useful_functions_genome for more
details.</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_TTS</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">TTSs</span><span class="p">[</span><span class="s2">&quot;RhoTerm&quot;</span><span class="p">][</span><span class="mi">2791688</span><span class="p">]</span><span class="o">.</span><span class="n">left</span>
<span class="go">2791688</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Genome.Genome.load_GO">
<span class="sig-name descname"><span class="pre">load_GO</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Genome.Genome.load_GO" title="Permalink to this definition"></a></dt>
<dd><p>load_GO loads file specified in GO.info to assign GO terms to genes.</p>
<dl class="simple">
<dt>Creates:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.GO (dict. of dict)</dt><dd><p>new attribute of the Genome instance. self.GO is a dictionary of
shape {annot_syst: {GOterm: list of genes}} i.e. one subdictionary
is created for each annotation system (such as GOc, COG or domain)
listed in GO.info file.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.genes[locus].GO (list)</dt><dd><p>new attribute of Gene instances related to the Genome instance given
as argument. List of terms (such as GO terms) associated to the gene.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The information importation requires a GO.info file,
containing column indices of each information in the data file and
some additional information, in the following order:
[0] Annotation system [1] Filename [2] Locus tag column
[3] GOterm column [4] Separator
GO.info and the data files have to be in the /GO/ directory</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Other annotation systems such as COG, or domain assignment can also
be used.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic annotation. If no annotation is
loaded, the load_annotation method with the default “sequence.gff3”
file is computed. To use another annotation, please load an
annotation before using this method.</p>
</div>
<dl>
<dt>Example</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_GO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">GO</span><span class="p">[</span><span class="s1">&#39;GO&#39;</span><span class="p">][</span><span class="s1">&#39;GO:0000100&#39;</span><span class="p">]</span>
<span class="go">[&#39;Dda3937_01944&#39;, &#39;Dda3937_03618&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="s2">&quot;Dda3937_00004&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">GO</span>
<span class="go">{&#39;GO&#39;: [&#39;GO:0005829&#39;,&#39;GO:0055114&#39;,&#39;GO:0046872&#39;,&#39;GO:0042802&#39;,</span>
<span class="go">        &#39;GO:0000166&#39;,&#39;GO:0006177&#39;,&#39;GO:0003938&#39;,&#39;GO:0005829&#39;,</span>
<span class="go">        &#39;GO:0055114&#39;,&#39;GO:0046872&#39;,&#39;GO:0042802&#39;,&#39;GO:0000166&#39;,</span>
<span class="go">        &#39;GO:0006177&#39;,&#39;GO:0003938&#39;]}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-useful_functions_genome">
<span id="useful-functions"></span><h2>useful functions<a class="headerlink" href="#module-useful_functions_genome" title="Permalink to this heading"></a></h2>
<p>Functions called by Genome methods</p>
<dl class="py function">
<dt class="sig sig-object py" id="useful_functions_genome.read_seq">
<span class="sig-prename descclassname"><span class="pre">useful_functions_genome.</span></span><span class="sig-name descname"><span class="pre">read_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#useful_functions_genome.read_seq" title="Permalink to this definition"></a></dt>
<dd><p>Called by load_seq, read_seq allows genomic sequence to be loaded from a
.fasta file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>filename</strong> (<em>str.</em>) – name of the file containing the DNA sequence in
FASTA format.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>genomic sequence (str.)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="useful_functions_genome.load_gff">
<span class="sig-prename descclassname"><span class="pre">useful_functions_genome.</span></span><span class="sig-name descname"><span class="pre">load_gff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annotations_filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#useful_functions_genome.load_gff" title="Permalink to this definition"></a></dt>
<dd><p>Called by load annotation, load_gff allows genes to be loaded from .gff
or .gff3 annotation file. For each gene of the annotation, onne Gene
object is created and intialized with the following attributes:
locus_tag, ID, name, left (left coordinate), right (right coordinate),
middle, length, strand, start, end and ASAP.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>filename</strong> (<em>str.</em>) – name of the annotation file.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>Dictionary of shape {locus: Gene object} with each Gene
object initialized with the following attributes:</p>
<blockquote>
<div><ul class="simple">
<li><p>locus_tag: “locus_tag” recorded in the annotation,</p></li>
<li><p>ID: “ID” or “locus_tag” if no “ID” is associated this gene in
the annotation,</p></li>
<li><p>name: “gene” or “locus_tag” if no “gene” is associated to this
gene in the annotation,</p></li>
<li><p>ASAP: “ASAP” name (By default: ‘’).</p></li>
<li><p>strand: gene strand,</p></li>
<li><p>left, right: gene coordinates (does not take into account the
strand, ie right &gt; left)</p></li>
<li><p>start, end, middle: positions of the beginning, the middle and
the end of the gene</p></li>
<li><p>length: gene length (=right-left)</p></li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary</p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>ASAP name  is found only if it is noted on the same line as the
“name” line or the line next to it</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="useful_functions_genome.load_annot_general">
<span class="sig-prename descclassname"><span class="pre">useful_functions_genome.</span></span><span class="sig-name descname"><span class="pre">load_annot_general</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annotations_filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separator</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tag_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ID_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ASAP_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_line</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#useful_functions_genome.load_annot_general" title="Permalink to this definition"></a></dt>
<dd><p>Called by load annotation, load_annot_general allows genes to be loaded
from any annotation file, although most of the time, annotation is loaded
from gff files (using the load_gff function). The annotation file has to
contains one row per genes and the following columns: locus_tag, ID,
name, ASAP, left coordinate, right coordinate and strand.
For each gene of the annotation, one Gene object is created and
intialized with the following attributes: locus_tag, ID, name, left
(left coordinate), right (right coordinate), middle, length, strand,
start, end and ASAP.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str.</em>) – name of the annotation file</p></li>
<li><p><strong>separator</strong> (<em>str.</em>) – file separator</p></li>
<li><p><strong>tag_col</strong> (<em>int.</em>) – index of the column containing the locus tags</p></li>
<li><p><strong>name_column</strong> (<em>int.</em>) – index of the column containing the names</p></li>
<li><p><strong>ID_column</strong> (<em>int.</em>) – index of the column containing the ID</p></li>
<li><p><strong>strand_column</strong> (<em>int.</em>) – index of the column containing the strand</p></li>
<li><p><strong>left_column</strong> (<em>int.</em>) – index of the column containing the left
coordinates</p></li>
<li><p><strong>right_column</strong> (<em>int.</em>) – index of the column containing the right
coordinates</p></li>
<li><p><strong>ASAP_column</strong> (<em>int.</em>) – index of the column containing the ASAP names</p></li>
<li><p><strong>start_line</strong> (<em>int.</em>) – file start line</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dict. of shape {locus: Gene object} with each Gene object
initialized with the following attributes: locus_tag, ID, name,
left (left coordinate), right (right coordinate), middle, length,
strand, start, end and ASAP.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Column numbering starts at 0.</p>
<p>Strands can be noted with one of the following writings:</p>
<blockquote>
<div><ul class="simple">
<li><p>Forward: “forward”,”1”,”+”,”true”,”plus” (case insensitive)</p></li>
<li><p>Reverse: “complement”,”-1”,”-“,”false”,”minus” (case insensitive)</p></li>
</ul>
</div></blockquote>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="useful_functions_genome.load_TSS_cond">
<span class="sig-prename descclassname"><span class="pre">useful_functions_genome.</span></span><span class="sig-name descname"><span class="pre">load_TSS_cond</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genes_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">TSS_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_line</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separator</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandcol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genescol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sigcol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sitescol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scorecol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#useful_functions_genome.load_TSS_cond" title="Permalink to this definition"></a></dt>
<dd><p>Called by load_TSS, load_TSS_cond allows TSS data to be loaded from any
file with one row per TSS and at least the following columns (separated
by a separator that is not a comma): TSS position and DNA strand.
Additional columns can be locus tags, sigma factors, sites positions and
score. The sites positions columns is divided in 4 part separated by
commas: -10 element left coordinate, -10 element right coordinate,
-35 element left coordinate and -35 element right coordinate.
For each TSS of the annotation, one TSS object is created and intialized
with the attributes available in the data file (at least pos and strand).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genes_dict</strong> (<em>dict.</em>) – dictionary of shape {locus tag: Gene object}</p></li>
<li><p><strong>filename</strong> (<em>str.</em>) – name of the annotation file</p></li>
<li><p><strong>TSS_column</strong> (<em>int.</em>) – index of the column containing the TSS positions</p></li>
<li><p><strong>start_line</strong> (<em>int.</em>) – file start line</p></li>
<li><p><strong>separator</strong> (<em>str.</em>) – file separator (other than commas !)</p></li>
<li><p><strong>strandcol</strong> (<em>int.</em>) – index of the column containing the strands in the
file</p></li>
<li><p><strong>genescol</strong> (<em>int.</em>) – index of the column containing the locus tags of the
genes associated to each TSS in the file. By default: None
(ie not in the file)</p></li>
<li><p><strong>sigcol</strong> (<em>int.</em>) – index of the column containing the name of the sigma
factor associated to each TSS in the file
By default: None (ie not on file)</p></li>
<li><p><strong>sitescol</strong> (<em>int.</em>) – index of the column containing the sites positions
(ie the coordinates of the -10 and the -35 elements)
associated to each TSS in the file
By default: None (ie not on file)</p></li>
<li><p><strong>scorecol</strong> (<em>int.</em>) – index of the TSS scores column in the file
By default: None (ie not on file)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>dict. of shape {TSS position: TSS object} with each TSS object
initialized with, at least, the following attributes: pos and strand.
Depending on the data present in the file and passed as arguments,
the following attributes may also have been added: genes, score and
promoter. The promoter attribute is a dictionary containing, for each
sigma factor (keys) a subdictionary (value). The first created key of
this subdictionary is “sites” and the associated value is a tuple
containing the positions of promoter elements (-10l,-10r,-35l,-35r)
with l = left coordinate and r = right coordinate.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Column numbering starts at 0.</p>
<p>Strands can be noted with one of the following writings:</p>
<blockquote>
<div><ul class="simple">
<li><p>Forward: “forward”,”1”,”+”,”true”,”plus” (case insensitive)</p></li>
<li><p>Reverse: “complement”,”-1”,”-“,”false”,”minus” (case insensitive)</p></li>
</ul>
</div></blockquote>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="useful_functions_genome.load_TTS_cond">
<span class="sig-prename descclassname"><span class="pre">useful_functions_genome.</span></span><span class="sig-name descname"><span class="pre">load_TTS_cond</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separator</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_line</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leftcol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rightcol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandcol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rhocol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqcol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scorecol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genescol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#useful_functions_genome.load_TTS_cond" title="Permalink to this definition"></a></dt>
<dd><p>Called by load_TTS, load_TTS_cond allows TTS data to be loaded from any
file with one row per TTS and at least the following columns:
left coordinate, right coordinate, strand and rho_dpdt. Additional
columns can be genes locus tags, TTS sequence and TTS score.
For each TTS of the annotation, one TTS object is created and intialized
with the attributes available in the data file (at least left, right,
rho_dpdt and strand).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str.</em>) – name of the annotation file</p></li>
<li><p><strong>separator</strong> (<em>str.</em>) – file separator</p></li>
<li><p><strong>start_line</strong> (<em>int.</em>) – file start line</p></li>
<li><p><strong>leftcol</strong> (<em>int.</em>) – index of the column containing the TTS left
coordinates</p></li>
<li><p><strong>rightcol</strong> (<em>int.</em>) – index of the column containing the TTS right
coordinates</p></li>
<li><p><strong>strandcol</strong> (<em>int.</em>) – index of the column containing the TTS strands</p></li>
<li><p><strong>rhocol</strong> (<em>int.</em>) – index of the column containing the information about
the rho dependency of the TTS</p></li>
<li><p><strong>seqcol</strong> (<em>int.</em>) – index of the column containing the TTS sequences
By default: None (ie not on file)</p></li>
<li><p><strong>scorecol</strong> (<em>int.</em>) – index of the TTS scores column in the file
By default: None (ie not on file)</p></li>
<li><p><strong>genescol</strong> (<em>int.</em>) – index of the column containing the locus tags of the
genes associated to each TTS in the file.
By default: None (ie not on file)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dict. of shape {TTS position: TTS object} with each TTS object
initialized with, at least, the following attributes: left,
right, start, end, strand and rho_dpdt. Depending on the data
present in the file and passed as arguments, the following attributes
may also have been added: genes, score and seq.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Column numbering starts at 0.</p>
<p>Strands can be noted with one of the following writings:</p>
<blockquote>
<div><ul class="simple">
<li><p>Forward: “forward”,”1”,”+”,”true”,”plus” (case insensitive)</p></li>
<li><p>Reverse: “complement”,”-1”,”-“,”false”,”minus” (case insensitive)</p></li>
</ul>
</div></blockquote>
<p>Rho dependent TTS can be noted with one of the following writings:
“True”,”1” (case insensitive)</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="useful_functions_genome.load_TU_cond">
<span class="sig-prename descclassname"><span class="pre">useful_functions_genome.</span></span><span class="sig-name descname"><span class="pre">load_TU_cond</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">startcol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">endcol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandcol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genescol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">startline</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separator</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#useful_functions_genome.load_TU_cond" title="Permalink to this definition"></a></dt>
<dd><p>Called by load_TU, load_TU_cond allows TU data to be loaded from any file
with one row per TU and the following columns: start position, end
position, strand and genes locus tags. For each TU of the annotation, one
TU object is created and intialized with the attributes start, end,
strand and genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str.</em>) – name of the annotation file</p></li>
<li><p><strong>startcol</strong> (<em>int.</em>) – index of the column containing the TU start positions</p></li>
<li><p><strong>endcol</strong> (<em>int.</em>) – index of the column containing the TU end positions</p></li>
<li><p><strong>strandcol</strong> (<em>int.</em>) – index of the column containing the TU strands</p></li>
<li><p><strong>genescol</strong> (<em>int.</em>) – index of the column containing the locus tags of the
genes associated to each TU in the file.
By default: None (ie not on file)</p></li>
<li><p><strong>start_line</strong> (<em>int.</em>) – file start line</p></li>
<li><p><strong>separator</strong> (<em>str.</em>) – file separator</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>dict. of shape {TU start: TU object} with each TU object
initialized with the attributes start, end, strand and genes.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Column numbering starts at 0.</p>
<p>Strands can be noted with one of the following writings:</p>
<blockquote>
<div><ul class="simple">
<li><p>Forward: “forward”,”1”,”+”,”true”,”plus” (case insensitive)</p></li>
<li><p>Reverse: “complement”,”-1”,”-“,”false”,”minus” (case insensitive)</p></li>
</ul>
</div></blockquote>
</div>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Presentation.html" class="btn btn-neutral float-left" title="Presentation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Transcriptome.html" class="btn btn-neutral float-right" title="Transcriptome" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Pineau, Forquet, Meyer.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>