<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>HiC &mdash; topo_package 1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="plot_genome" href="plot_genome.html" />
    <link rel="prev" title="Chipseq" href="Chipseq.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            topo_package
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Presentation.html">Presentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="Genome.html">Genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="Transcriptome.html">Transcriptome</a></li>
<li class="toctree-l1"><a class="reference internal" href="Chipseq.html">Chipseq</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">HiC</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#class-methods">class methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-useful_functions_HiC">useful functions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="plot_genome.html">plot_genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="GO.html">GO</a></li>
<li class="toctree-l1"><a class="reference internal" href="stat_analysis.html">statistical analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="TSS_TTS_TU.html">TSS_TTS_TU</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">topo_package</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">HiC</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/HiC.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="hic">
<h1>HiC<a class="headerlink" href="#hic" title="Permalink to this heading"></a></h1>
<section id="class-methods">
<h2>class methods<a class="headerlink" href="#class-methods" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="HiC.HiC">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">HiC.</span></span><span class="sig-name descname"><span class="pre">HiC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#HiC.HiC" title="Permalink to this definition"></a></dt>
<dd><p>The HiC class is used to gather spatial information obtained from HiC data
analysis with tools such as Chromosight. The two patterns analyzed here are
borders (between two topological domains) and loops (more specifically the
positions of the two regions that come into contact).</p>
<p>Each HiC instance has to be initialized with an organism name</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">hc</span> <span class="o">=</span> <span class="n">HiC</span><span class="o">.</span><span class="n">HiC</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="HiC.HiC.load_hic_borders">
<span class="sig-name descname"><span class="pre">load_hic_borders</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#HiC.HiC.load_hic_borders" title="Permalink to this definition"></a></dt>
<dd><p>load_hic_borders imports a list of borders from a data file (typically
a borders.tsv data file obtained with Chromosight) containing at least,
for each border, its position (bin number) and optionally its identification
score (pearson correlation coefficient between the border kernel and the
detected pattern), pvalue and qvalue.
Creates 2 new attributes of the HiC instance:</p>
<blockquote>
<div><ul class="simple">
<li><dl class="simple">
<dt>self.borders (dict. of dict.)</dt><dd><p>dictionary containing one subdictionary per condition
with the shape self.borders[cond]={bin_nb: {“score”:
score, “pval”: pvalue, “qval”: qvalue,”binsize”:binsize}}.
N.B.: If data are binned at 2kb, the bin with number 10
corresponds to data between 20000 and 22000.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.borders_pos (dict. of dict)</dt><dd><p>dictionary containing one subdictionary per
condition, with the shape self.borders_pos[cond] =
{‘borders’:list of positions that are in a border
‘no_borders’:list of positions that are not in a border}</p>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>cond</strong> (<em>Optional</em><em> [</em><em>list of str.</em><em>]</em>) – selection of one or several
conditions (1 condition corresponds to 1 data file and
each condition has to be listed in the border.info file).
By default: cond =’all’ ie all available conditions are
loaded.</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The data importation requires a borders.info file that contains the
column indices of each information in the data file and some
additional information, in the following order:</p>
<blockquote>
<div><ul class="simple">
<li><p>(required) [0] Condition, [1] Filename, [2] Startline,
[3] Separator, [4] Bin, [5] Binsize (in b)</p></li>
<li><p>(optional) [6] Score, [7] Pvalue,[8] Qvalue</p></li>
</ul>
</div></blockquote>
<p>borders.info and data file have to be in the /HiC/Borders/ directory</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>See Chromosight documentation for more details about the input format.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Make sure that the version of the genome is the same as
the version used in the HiC analysis workflow</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">hc</span> <span class="o">=</span> <span class="n">HiC</span><span class="o">.</span><span class="n">HiC</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hc</span><span class="o">.</span><span class="n">load_hic_borders</span><span class="p">(</span><span class="s2">&quot;WT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hc</span><span class="o">.</span><span class="n">borders_pos</span><span class="p">[</span><span class="s1">&#39;WT&#39;</span><span class="p">][</span><span class="s1">&#39;no_borders&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">]</span>
<span class="go">[2001,2002,2003,2004,2005]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiC.HiC.load_hic_loops">
<span class="sig-name descname"><span class="pre">load_hic_loops</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#HiC.HiC.load_hic_loops" title="Permalink to this definition"></a></dt>
<dd><p>load_hic_loops imports a list of loops from a data file (typically
a loops.tsv data file obtained with Chromosight) containing, for each
loop, at least the loop coordinates (2 bin numbers) and optionally the
loop score (pearson correlation coefficient between the loop kernel and
the detected pattern), the p-value and the q-value.</p>
<dl class="simple">
<dt>Creates 2 new attributes of the HiC instance:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.loops (dict. of dict.)</dt><dd><p>dictionary containing one subdictionary per condition with
the shape self.loops[cond] =
{(bin1_nb,bin2_nb): {“score”: score, “pval”: pvalue,
“qval”: qvalue, “binsize”:binsize}}
N.B.: If data are binned at 2kb, the bin with number 10
corresponds to data between 20000 and 22000.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.loops_pos (dict. of dict)</dt><dd><p>dictionary containing one subdictionary per condition, with
the shape self.loops_pos[cond] =
{‘loops’:list of positions that are in a loop,
‘no_loops’:list of positions that are not in a loop}</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>self</strong> (<em>HiC instance</em>) – </p></li>
<li><p><strong>cond</strong> (<em>Optional</em><em> [</em><em>list of str.</em><em>]</em>) – selection of one or several
conditions (1 condition corresponds to 1 data file and
has to be in the loops.info file). By default:
cond =’all’ ie all available data are loaded.</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The data importation requires a loops.info file that contains the
column indices of each information in the data file and some
additional information, in the following
order:</p>
<blockquote>
<div><ul class="simple">
<li><p>(required) [0] Condition, [1] Filename, [2] Startline,
[3] Separator, [4] Bin1, [5] Bin2, [6] Binsize (in b),</p></li>
<li><p>(optional) [7] Score, [8] Pvalue,[9] Qvalue</p></li>
</ul>
</div></blockquote>
<p>loops.info and data file have to be in the /HiC/Loops/ directory</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>See Chromosight documentation for more details about the input format.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Make sure that the version of the genome is the same as
the version used in the HiC analysis workflow</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">hc</span> <span class="o">=</span> <span class="n">HiC</span><span class="o">.</span><span class="n">HiC</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hc</span><span class="o">.</span><span class="n">load_hic_loops</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hc</span><span class="o">.</span><span class="n">loops_pos</span><span class="p">[</span><span class="s1">&#39;WT&#39;</span><span class="p">][</span><span class="s1">&#39;no_borders&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">]</span>
<span class="go">[2001,2002,2003,2004,2005]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hc</span><span class="o">.</span><span class="n">loops</span><span class="p">[</span><span class="s1">&#39;WT1&#39;</span><span class="p">]</span>
<span class="go">{(42000, 80000): {&#39;binsize&#39;: 2000},</span>
<span class="go"> (238000, 252000): {&#39;binsize&#39;: 2000}</span>
<span class="go"> ...}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiC.HiC.load_loops_genes">
<span class="sig-name descname"><span class="pre">load_loops_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#HiC.HiC.load_loops_genes" title="Permalink to this definition"></a></dt>
<dd><p>load_loops_genes imports a list of loops determined with HiC from a
data file (typically a loops.tsv data file obtained with Chromosight)
containing at least the loops coordinates (2 bin numbers) and
determines the list of genes overlapping one of the loops positions.</p>
<dl class="simple">
<dt>Creates:</dt><dd><dl class="simple">
<dt>self.loops_pos (dict. of dict.)</dt><dd><p>new attribute of the Genome
instance. Dictionary containing one dictionary per
condition listed in loops.info. Each subdictionary has
2 keys: “loops” and “no_loops” and contains the
corresponding lists of genomic positions.</p>
</dd>
<dt>self.loops_genes (dict. of dict.)</dt><dd><p>New attribute of the Genome instance.
Dictionary containing one dictionary per condition listed
in loops.info. The key of this dictionary contains the
condition name and the chosen window size (for example
“cond_w100b” for a 100b window).
Each subdictionary has 2 keys: “loops” and “no_loops” and
contains the corresponding lists of genes.
For example, self.loops_genes[cond_w0b][“loops”] returns
the list of genes that overlap the loops positions.</p>
</dd>
<dt>self.genes[locus].is_loop (dict.)</dt><dd><p>New attribute of Gene instances related to the Genome
instance given as argument.Dictionary of shape
{condition:boolean}. The boolean is True if the gene
overlaps the loop positions (window included).</p>
</dd>
</dl>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cond</strong> (<em>Optional</em><em> [</em><em>list of str.</em><em>]</em>) – selection of one or several
conditions (1 condition corresponds to 1 data file and
has to be in the loops.info file). By default:
cond =’all’ ie all available data are loaded.</p></li>
<li><p><strong>per_genes</strong> (<em>Optional</em><em> [</em><em>Bool.</em><em>]</em>) – if True, determines the list of
genes overlapping the loops positions and returns the
outputs self.loops_genes and self.genes[locus].is_loop
described below.</p></li>
<li><p><strong>window</strong> (<em>Optional</em><em> [</em><em>int.</em><em>]</em>) – <p>window around the loop positions (ie
around the 2 bins) for the seeking of overlapping genes
(Default: 0). All genes overlaping any position between:</p>
<blockquote>
<div><ul>
<li><p>loops_start_bin1 - window and loops_end_bin1 + window or</p></li>
<li><p>loops_start_bin2 - window and loops_end_bin2 + window</p></li>
</ul>
</div></blockquote>
<p>are considered “loops” genes.</p>
</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The data importation requires a loops.info file that contains
the column indices of each information in the data file and
some additional information, in the following order:</p>
<blockquote>
<div><ul class="simple">
<li><p>(required) [0] Condition, [1] Filename, [2] Startline,
[3] Separator, [4] Bin1, [5] Bin2, [6] Binsize (in b),</p></li>
<li><p>(optional) [7] Score, [8] Pvalue,[9] Qvalue</p></li>
</ul>
</div></blockquote>
<p>loops.info and data file have to be in the /HiC/Loops/ directory</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The position (pos) is the bin number. If data are binned at 2kb,
the bin with number 10 corresponds to data between 20000 and 22000.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic sequence and a genomic annotation.
If no annotation is loaded, the load_annotation method with the
default “sequence.gff3” file is computed. If no sequence is loaded,
the load_seq method with de defualt “sequence.fasta” is computed.
To use another annotation or sequence, please load them to your
Transcriptome instance with the following commands before using this
method:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">Genome</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">HiC</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">HC</span> <span class="o">=</span> <span class="n">HiC</span><span class="o">.</span><span class="n">HiC</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="n">HC</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_annotation</span><span class="p">(</span><span class="n">annot_file</span><span class="o">=</span><span class="n">chosen_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_seq</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">chosen_file2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">HC</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">genes</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">HC</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">length</span>
</pre></div>
</div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Make sure that the version of the genome is the same as
the version used in the HiC analysis workflow</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_loops_genes</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">loops_pos</span><span class="p">[</span><span class="s1">&#39;WT2kb_bin2000b&#39;</span><span class="p">][</span><span class="s1">&#39;loops&#39;</span><span class="p">]</span>
<span class="go">[254000,254001,254002,254003,...]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">loops_genes</span><span class="p">[</span><span class="s1">&#39;WT2kb_bin2000b_w0b&#39;</span><span class="p">][</span><span class="s1">&#39;loops&#39;</span><span class="p">]</span>
<span class="go">[&#39;Dda3937_00221&#39;,&#39;Dda3937_04438&#39;,&#39;Dda3937_03673&#39;,...]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="s1">&#39;Dda3937_00221&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">is_loop</span>
<span class="go">{&#39;acid2kb_bin2000b&#39;: False, &#39;WT2kb_bin2000b&#39;: True}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="HiC.HiC.load_borders_genes">
<span class="sig-name descname"><span class="pre">load_borders_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#HiC.HiC.load_borders_genes" title="Permalink to this definition"></a></dt>
<dd><p>load_hic_borders_genes imports a list of borders from a data file
(typically a borders.tsv data file obtained with Chromosight)
containing at least the border position (bin number) and determines
the list of genes overlapping one of the borders.</p>
<dl class="simple">
<dt>Creates:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.borders_genes (dict. of dict.)</dt><dd><p>created only if per_genes = True. New attribute of the
Genome instance. Dictionary containing one dictionary
per condition listed in borders.info. The key of this
dictionary contains the condition name and the chosen
window size (for example “cond_w100b” for a 100b window).
Each subdictionary has 2 keys: “borders” and “no_borders”
and contains the corresponding lists of genes. For
example, self.borders_genes[cond_w0b][“borders”] returns
the list of genes that overlap the borders positions.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.borders_pos (dict. of dict)</dt><dd><p>new attribute of the HiC instance, dictionary containing
one subdictionary per condition, with the shape
self.borders_pos[cond] =
{‘borders’:list of positions that are in a border,
‘no_borders’:list of positions that are not in a border}</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.genes[locus].is_border (dict.)</dt><dd><p>created only if per_genes = True.
New attribute of Gene instances related to the Genome
instance given as argument.Dictionary of shape
{condition:boolean}. The boolean is True if the gene
overlaps the border positions (window included).</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cond</strong> (<em>Optional</em><em> [</em><em>list of str.</em><em>]</em>) – selection of one or several
conditions (1 condition corresponds to 1 data file and
has to be in the borders.info file). By default:
cond =’all’ ie all available data are loaded.</p></li>
<li><p><strong>per_genes</strong> (<em>Optional</em><em> [</em><em>Bool.</em><em>]</em>) – if True, determines the list of
genes overlapping the borders positions and returns the
outputs self.borders_genes and self.genes[locus].is_border
described below.</p></li>
<li><p><strong>window</strong> (<em>Optional</em><em> [</em><em>int.</em><em>]</em>) – window around the border positions ie
around the bin) for the seeking of overlapping genes
(Default: 0). All genes overlaping any position between
border_start_bin - window and border_end_bin + window
are considered “borders” genes.</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The data importation requires a borders.info file that contains
the column indices of each information in the data file and some
additional information, in the following order:</p>
<blockquote>
<div><ul class="simple">
<li><p>(required) [0] Condition, [1] Filename, [2] Startline,
[3] Separator, [4] Bin, [5] Binsize (in b)</p></li>
<li><p>(optional) [6] Score, [7] Pvalue,[8] Qvalue</p></li>
</ul>
</div></blockquote>
<p>borders.info and data file have to be in the /HiC/Borders/ directory</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The position (pos) is the bin number. If data are binned at 2kb,
the bin with number 10 corresponds to data between 20000 and 22000.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic sequence and a genomic annotation.
If no annotation is loaded, the load_annotation method with the
default “sequence.gff3” file is computed. If no sequence is loaded,
the load_seq method with de defualt “sequence.fasta” is computed.
To use another annotation or sequence, please load them to your
Transcriptome instance with the following commands before using this
method:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">Genome</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">HiC</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">HC</span> <span class="o">=</span> <span class="n">HiC</span><span class="o">.</span><span class="n">HiC</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="n">HC</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_annotation</span><span class="p">(</span><span class="n">annot_file</span><span class="o">=</span><span class="n">chosen_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_seq</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">chosen_file2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">HC</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">genes</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">HC</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">length</span>
</pre></div>
</div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>make sure that the version of the genome is the same as
the version used in the HiC analysis workflow</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_borders_genes</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">borders_pos</span><span class="p">[</span><span class="s2">&quot;WT2_2kb_bin2000b&quot;</span><span class="p">][</span><span class="s1">&#39;borders&#39;</span><span class="p">]</span>
<span class="go">[20000,20001,20002,20003,20004,...]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">borders_genes</span><span class="p">[</span><span class="s1">&#39;acid1kb_bin1000b_w0b&#39;</span><span class="p">][</span><span class="s1">&#39;borders&#39;</span><span class="p">]</span>
<span class="go">[&#39;Dda3937_00158&#39;,&#39;Dda3937_01107&#39;,&#39;Dda3937_01108&#39;,...]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="s1">&#39;Dda3937_00221&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">is_border</span>
<span class="go">{&#39;acid1kb_bin1000b&#39;: False,&#39;WT2_2kb_bin2000b&#39;: False}</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</section>
<section id="module-useful_functions_HiC">
<span id="useful-functions"></span><h2>useful functions<a class="headerlink" href="#module-useful_functions_HiC" title="Permalink to this heading"></a></h2>
<p>Functions called by HiC methods</p>
<dl class="py function">
<dt class="sig sig-object py" id="useful_functions_HiC.load_HiC_site_cond">
<span class="sig-prename descclassname"><span class="pre">useful_functions_HiC.</span></span><span class="sig-name descname"><span class="pre">load_HiC_site_cond</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">site_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path2file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">startline</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin1_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binsize</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin2_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pval_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qval_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#useful_functions_HiC.load_HiC_site_cond" title="Permalink to this definition"></a></dt>
<dd><p>Called by load_HiC_borders and load_HiC_loops, allows borders and loops
data to be loaded by specifying files, and where each information is (one
column for each type of information).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>type</strong> (<em>str.</em>) – HiC detected motif type (“borders” or “loops”)</p></li>
<li><p><strong>path2file</strong> (<em>str.</em>) – path to the file containing the data</p></li>
<li><p><strong>startline</strong> (<em>int.</em>) – file start line</p></li>
<li><p><strong>sep</strong> (<em>str.</em>) – file separator</p></li>
<li><p><strong>bin1_col</strong> (<em>int.</em>) – index of the column containing the position of the
first bin</p></li>
<li><p><strong>binsize</strong> (<em>int.</em>) – binsize (in b)</p></li>
<li><p><strong>bin2_col</strong> (<em>Optional</em><em> [</em><em>int.</em><em>]</em>) – index of the column containing the position of the
second bin</p></li>
<li><p><strong>score_col</strong> (<em>Optional</em><em> [</em><em>int.</em><em>]</em>) – index of the column containing the position of the score</p></li>
<li><p><strong>pval_col</strong> (<em>Optional</em><em> [</em><em>int.</em><em>]</em>) – index of the column containing the position of the pvalue</p></li>
<li><p><strong>qval_col</strong> (<em>Optional</em><em> [</em><em>int.</em><em>]</em>) – index of the column containing the position of the qvalue</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Column numbering starts at 0.</p>
</div>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Chipseq.html" class="btn btn-neutral float-left" title="Chipseq" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="plot_genome.html" class="btn btn-neutral float-right" title="plot_genome" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Pineau, Forquet, Meyer.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>