<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>plot_genome &mdash; topo_package 1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="GO" href="GO.html" />
    <link rel="prev" title="HiC" href="HiC.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            topo_package
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Presentation.html">Presentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="Genome.html">Genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="Transcriptome.html">Transcriptome</a></li>
<li class="toctree-l1"><a class="reference internal" href="Chipseq.html">Chipseq</a></li>
<li class="toctree-l1"><a class="reference internal" href="HiC.html">HiC</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">plot_genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="GO.html">GO</a></li>
<li class="toctree-l1"><a class="reference internal" href="stat_analysis.html">statistical analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="TSS_TTS_TU.html">TSS_TTS_TU</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">topo_package</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">plot_genome</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/plot_genome.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-plot_genome">
<span id="plot-genome"></span><h1>plot_genome<a class="headerlink" href="#module-plot_genome" title="Permalink to this heading"></a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="plot_genome.plot_region">
<span class="sig-prename descclassname"><span class="pre">plot_genome.</span></span><span class="sig-name descname"><span class="pre">plot_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gen</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">beg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RNASeq_cond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">signals_cond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/maiwenn/these/resdir/Genome_plot/'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Genome_2023-04-25</span> <span class="pre">11:42:58.662478'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_extension</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'.pdf'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#plot_genome.plot_region" title="Permalink to this definition"></a></dt>
<dd><p>Plots experimental information (one subplot per experiment) and gene
annotations for a region. The genes annotation subplot is created with
the subplot_genes function. RNASeq subplots and Chipseq subplots
are created with the subplot_rnaseq_coverage and subplot_signal functions
respectively.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gen</strong> – Genome instance</p></li>
<li><p><strong>beg</strong> (<em>int.</em>) – beginning of the region to be plotted</p></li>
<li><p><strong>end</strong> (<em>int.</em>) – end of the region to be plotted</p></li>
<li><p><strong>RNASeq_cond</strong> (<em>list of str.</em>) – list of the RNASeq conditions names to
represent. Left empty by default ie no cover is plotted.</p></li>
<li><p><strong>tr_object</strong> (<em>Transcriptome instance</em>) – Transcriptome instance with
loaded RNASeq coverages</p></li>
<li><p><strong>signals_cond</strong> (<em>list of str.</em>) – list of the Chipseq conditions names to
represent. Left empty by default ie no signal is plotted.</p></li>
<li><p><strong>ch_object</strong> (<em>Chipseq instance</em>) – Chipseq instance with loaded signals.</p></li>
<li><p><strong>gene_names</strong> (<em>Optional</em><em> [</em><em>Bool.</em><em>]</em>) – If true, the name of the genes is noted on their
representation. (True by default).</p></li>
<li><p><strong>output_dir</strong> (<em>Optional</em><em> [</em><em>str.</em><em>]</em>) – output directory</p></li>
<li><p><strong>output_file</strong> (<em>Optional</em><em> [</em><em>str.</em><em>]</em>) – output filename for plot</p></li>
<li><p><strong>file_extension</strong> (<em>Optional</em><em> [</em><em>str.</em><em>]</em>) – Graphic file extension type (.pdf by default)</p></li>
<li><p><strong>R_ylabels</strong> (<em>Optional</em><em> [</em><em>list of str.</em><em>]</em>) – y-axis label for RNASeq data subplots.
By default, conditions names given in input with the RNASeq_cond
argument.</p></li>
<li><p><strong>S_ylabels</strong> (<em>Optional</em><em> [</em><em>list of str.</em><em>]</em>) – y-axis label for RNASeq data subplots.
By default, conditions names given in input with the signals_cond
argument.</p></li>
<li><p><strong>vlines</strong> (<em>Optional</em><em> [</em><em>dict.</em><em>]</em>) – Dictionnary of shape {position: annotation text}.
One green vertical line and its annotation are plotted for each
item of this dictionnary.</p></li>
<li><p><strong>figsize</strong> (<em>Optional</em><em> [</em><em>(</em><em>float</em><em>,</em><em>float</em><em>)</em><em>]</em>) – width and height in inches
(by default: (3.5,n) with n the number of suplots)</p></li>
<li><p><strong>hratios</strong> (<em>Optional</em><em> [</em><em>array-like</em><em>]</em>) – Relative heights of the subplots. If not given,
all rows will have the same height except the genes subplot.</p></li>
<li><p><strong>hspace</strong> (<em>Optional</em><em> [</em><em>float</em><em>]</em>) – Height space between subplots, expressed as a
fraction of the average axis height. By default: 0.05</p></li>
<li><p><strong>figpos</strong> (<em>Optional</em><em> [</em><em>array-like of length 4</em><em>]</em>) – Position of the subplots as a
fraction of figure width or height. The array contains the
positions in the following order: bottom, top, left and right.</p></li>
</ul>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic annotation. If no annotation is
loaded to the Genome instance, the load_annotation method of with
the default “sequence.gff3” file is computed. To use another
annotation, please load an annotation to your Genome instance before
using this method.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>To plot a ChIPseq coverage, all the selected conditions must first be loaded
on the Chipseq instance using one of the ChipSeq method (for example
load_signal or load_signals_average). This Chipseq instance must
then be given as input with the argument “ch_object”.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>To plot a RNAseq coverage, all the selected conditions must first be loaded
on the Transcriptome instance using tr.load_rnaseq_cov() and this
instance must be given as input with the argument “tr_object”.</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_annotation</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span> <span class="o">=</span> <span class="n">Transcriptome</span><span class="o">.</span><span class="n">Transcriptome</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">load_rnaseq_cov</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ch</span> <span class="o">=</span> <span class="n">Chipseq</span><span class="o">.</span><span class="n">Chipseq</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ch</span><span class="o">.</span><span class="n">load_signals_average</span><span class="p">(</span><span class="n">list_cond</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cond1&quot;</span><span class="p">,</span><span class="s2">&quot;cond2&quot;</span><span class="p">],</span>
<span class="gp">... </span>                        <span class="n">average_name</span><span class="o">=</span><span class="s2">&quot;cond12&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plot_genome</span><span class="o">.</span><span class="n">plot_region</span><span class="p">(</span><span class="n">g</span><span class="p">,</span><span class="mi">2101000</span><span class="p">,</span><span class="mi">2106000</span><span class="p">,</span>
<span class="go">                            RNASeq_cond = [&quot;WT&quot;],</span>
<span class="go">                            signals_cond=[&quot;cond12&quot;],</span>
<span class="go">                            R_ylabels=[&quot;WT&quot;],</span>
<span class="go">                            S_ylabels=[&quot;cond12&quot;],</span>
<span class="go">                            tr_object=tr,</span>
<span class="go">                            ch_object=ch,</span>
<span class="go">                            hratios = [1,2,2,1],</span>
<span class="go">                            figsize=(6,5),</span>
<span class="go">                            figpos=[0.1,0.98,0.15,0.95])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="plot_genome.subplot_genes">
<span class="sig-prename descclassname"><span class="pre">plot_genome.</span></span><span class="sig-name descname"><span class="pre">subplot_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gen</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">beg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xticks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#plot_genome.subplot_genes" title="Permalink to this definition"></a></dt>
<dd><p>Plots a subplot with the annoted oriented genes that are located in
the selected region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ax</strong> (<em>matplotlib Axes instance</em>) – </p></li>
<li><p><strong>gen</strong> (<em>Genome instance</em>) – </p></li>
<li><p><strong>beg</strong> (<em>int.</em>) – beginning of the region to be plotted</p></li>
<li><p><strong>end</strong> (<em>int.</em>) – end of the region to be plotted</p></li>
<li><p><strong>xticks</strong> (<em>list of float</em>) – xaxis’ tick locations.
None by default ie. the xaxis is left empty.</p></li>
<li><p><strong>gene_names</strong> (<em>Bool.</em>) – If true, the name of the genes is noted on their
representation. (True by default).</p></li>
</ul>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic annotation. If no annotation is
loaded to the Genome instance, the load_annotation method of with
the default “sequence.gff3” file is computed. To use another
annotation, please load an annotation to your Genome instance before using this method.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="plot_genome.subplot_rnaseq_coverage">
<span class="sig-prename descclassname"><span class="pre">plot_genome.</span></span><span class="sig-name descname"><span class="pre">subplot_rnaseq_coverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cond</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">beg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xticks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#plot_genome.subplot_rnaseq_coverage" title="Permalink to this definition"></a></dt>
<dd><p>Plots a subplot with the with the experimental data (RNASeq coverages on
both DNA strands) of the selected region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ax</strong> (<em>matplotlib Axes instance</em>) – </p></li>
<li><p><strong>tr</strong> (<em>Transcriptome instance</em>) – Transcriptome instance with loaded
RNASeq coverages</p></li>
<li><p><strong>cond</strong> (<em>str.</em>) – RNASeq condition name to represent.
WARNING: This condition must first be loaded on the Transcriptome
instance using tr.load_rnaseq_cov().</p></li>
<li><p><strong>beg</strong> (<em>int.</em>) – beginning of the region to be plotted</p></li>
<li><p><strong>end</strong> (<em>int.</em>) – end of the region to be plotted</p></li>
<li><p><strong>ylabel</strong> (<em>str.</em>) – y-axis label. By default, the condition name given
in input with the cond argument.</p></li>
<li><p><strong>xticks</strong> (<em>list of float</em>) – xaxis’ tick locations.
None by default ie. the xaxis is left empty.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="plot_genome.subplot_signal">
<span class="sig-prename descclassname"><span class="pre">plot_genome.</span></span><span class="sig-name descname"><span class="pre">subplot_signal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ch</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cond</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">beg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xticks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#plot_genome.subplot_signal" title="Permalink to this definition"></a></dt>
<dd><p>Plots a subplot with the with the experimental data (Chipseq signal) of
the selected region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ax</strong> (<em>matplotlib Axes instance</em>) – </p></li>
<li><p><strong>ch</strong> (<em>Chipseq instance</em>) – Chipseq instance with loaded signals.</p></li>
<li><p><strong>cond</strong> (<em>str.</em>) – Chipseq condition name to represent. WARNING: This
condition must first be loaded on the Chipseq instance using one
of the ChipSeq method (for example load_signal or
load_signals_average).</p></li>
<li><p><strong>beg</strong> (<em>int.</em>) – beginning of the region to be plotted</p></li>
<li><p><strong>end</strong> (<em>int.</em>) – end of the region to be plotted</p></li>
<li><p><strong>ylabel</strong> (<em>str.</em>) – y-axis label. By default, the condition name given
in input with the cond argument.</p></li>
<li><p><strong>xticks</strong> (<em>list of float</em>) – xaxis’ tick locations.
None by default ie. the xaxis is left empty.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="HiC.html" class="btn btn-neutral float-left" title="HiC" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="GO.html" class="btn btn-neutral float-right" title="GO" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Pineau, Forquet, Meyer.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>