<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Transcriptome &mdash; topo_package 1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Chipseq" href="Chipseq.html" />
    <link rel="prev" title="Genome" href="Genome.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            topo_package
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Presentation.html">Presentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="Genome.html">Genome</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Transcriptome</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#class-methods">class methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-useful_functions_transcriptome">useful functions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Chipseq.html">Chipseq</a></li>
<li class="toctree-l1"><a class="reference internal" href="HiC.html">HiC</a></li>
<li class="toctree-l1"><a class="reference internal" href="plot_genome.html">plot_genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="GO.html">GO</a></li>
<li class="toctree-l1"><a class="reference internal" href="stat_analysis.html">statistical analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="TSS_TTS_TU.html">TSS_TTS_TU</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">topo_package</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Transcriptome</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/Transcriptome.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="transcriptome">
<h1>Transcriptome<a class="headerlink" href="#transcriptome" title="Permalink to this heading"></a></h1>
<section id="class-methods">
<h2>class methods<a class="headerlink" href="#class-methods" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="Transcriptome.Transcriptome">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Transcriptome.</span></span><span class="sig-name descname"><span class="pre">Transcriptome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Transcriptome.Transcriptome" title="Permalink to this definition"></a></dt>
<dd><p>The Transcriptome class is the key class for loading expression data and
differential expression data for analysis and comparison. Typical data are
data obtained with high throughput methods such as RNASeq and processed with
common tools such as DeSeq2.</p>
<p>Each Transcriptome instance has to be initialized with an organism name</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span> <span class="o">=</span> <span class="n">Transcriptome</span><span class="o">.</span><span class="n">Transcriptome</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Transcriptome.Transcriptome.load_expression">
<span class="sig-name descname"><span class="pre">load_expression</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Transcriptome.Transcriptome.load_expression" title="Permalink to this definition"></a></dt>
<dd><p>loads expression data (such as log2rpkm) from files that are in the
/expression/ directory.</p>
<dl class="simple">
<dt>Creates:</dt><dd><dl class="simple">
<dt>self.genes[locus].expression (dict.): new attribute of Gene</dt><dd><p>instances related to the Transcriptome instance given as
argument. Dictionary of shape
{condition: expression level (float.)}</p>
</dd>
<dt>self.genes_valid_expr(dict.): Dictionary of shape</dt><dd><p>{condition: list of genes having an expression value}</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The data importation requires an expression.info file, containing
column indices of each information in the data file and some
additional information, in the following order:
[0] Condition [1] Filename [2] Locus_tag column
[3] Expression column [4] is Log ? (boolean) [5] Separator</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic annotation. If no annotation is
loaded, the load_annotation method with the default “sequence.gff3”
file is computed. To use another annotation, please load an
annotation to your Transcriptome instance with the following commands
before using this method:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">Genome</span><span class="o">,</span> <span class="nn">Transcriptome</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span> <span class="o">=</span> <span class="n">Transcriptome</span><span class="o">.</span><span class="n">Transcriptome</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="n">tr</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_annotation</span><span class="p">(</span><span class="n">annot_file</span><span class="o">=</span><span class="n">chosen_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">genes</span>
</pre></div>
</div>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span> <span class="o">=</span> <span class="n">Transcriptome</span><span class="o">.</span><span class="n">Transcriptome</span><span class="p">(</span><span class="s2">&quot;dickeya&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">load_expression</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="s2">&quot;Dda3937_00004&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">expression</span>
<span class="go">{&#39;WT_nov_stat(E10)_rpkm&#39;: 66.6413789332929,</span>
<span class="go">&#39;WT_stat(E3)&#39;: 6.980245227157392,</span>
<span class="go">&#39;WT_PGA_stat(E13)_rpkm&#39;: 13.9428053948966}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Transcriptome.Transcriptome.compute_fc_from_expr">
<span class="sig-name descname"><span class="pre">compute_fc_from_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ctrls</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Transcriptome.Transcriptome.compute_fc_from_expr" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Compute_fc_from_expr computes, for each gene:</dt><dd><ul class="simple">
<li><p>the mean of expression values (such as log2rpkm) in a list of
controls replicates (ctrls)</p></li>
<li><p>the mean of expression values in a list of tests replicates (conds)</p></li>
<li><p>the fold-change ie the difference between the mean control
and the mean test values</p></li>
<li><p>the pvalue of the Student-test for these two means</p></li>
</ul>
</dd>
<dt>Creates:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.genes[locus].fc_pval[condname] (tuple)</dt><dd><p>new attribute of Gene instances related to the
Transcriptome instance given as argument.Tuple of shape
(fold-change,pvalue).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.genes_valid_fc[condname] (list.)</dt><dd><p>list of genes having a fold-change and a pvalue corresponding
to the new condition</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>conds</strong> (<em>list of str.</em>) – list of expression conditions to use as test</p></li>
<li><p><strong>ctrls</strong> (<em>list of str.</em>) – list of expression conditions to use as control</p></li>
<li><p><strong>condname</strong> (<em>str.</em>) – condition name to use for the new fold-changes
and p-values</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The t-test is computed with scipy.stats.ttest_ind</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic annotation. If no annotation is
loaded, the load_annotation method with the default “sequence.gff3”
file is computed. To use another annotation, please load an
annotation to your Transcriptome instance with the following commands
before using this method:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">Genome</span><span class="o">,</span> <span class="nn">Transcriptome</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span> <span class="o">=</span> <span class="n">Transcriptome</span><span class="o">.</span><span class="n">Transcriptome</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="n">tr</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_annotation</span><span class="p">(</span><span class="n">annot_file</span><span class="o">=</span><span class="n">chosen_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">genes</span>
</pre></div>
</div>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span> <span class="o">=</span> <span class="n">Transcriptome</span><span class="o">.</span><span class="n">Transcriptome</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">compute_fc_from_expr</span><span class="p">([</span><span class="s2">&quot;Ctrl1&quot;</span><span class="p">,</span><span class="s2">&quot;Ctrl2&quot;</span><span class="p">],</span>
<span class="go">                            [&quot;Test1&quot;,&quot;Test2&quot;],</span>
<span class="go">                            [&quot;FC_Test12_Ctrl12&quot;])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="s2">&quot;b0002&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">fc_pval</span><span class="p">[</span><span class="s2">&quot;FC_Test12_Ctrl12&quot;</span><span class="p">]</span>
<span class="go">(-1.786, 0.003125)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Transcriptome.Transcriptome.load_fc_pval">
<span class="sig-name descname"><span class="pre">load_fc_pval</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Transcriptome.Transcriptome.load_fc_pval" title="Permalink to this definition"></a></dt>
<dd><p>loads Fold-changes and p-values (if available) from data files that
are in the /fold_changes/ directory.</p>
<dl class="simple">
<dt>Creates:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.genes[locus].fc_pval (dict.): new attribute of Gene instances</dt><dd><p>related to the Transcriptome instance given as argument.
Dictionary of shape {condition: (fold-change,pvalue)}.
If no p-value is available in the data file, 0 is set as
p-value.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.genes_valid_fc (dict.): Dictionary of shape</dt><dd><p>{condition: list of genes having at least a fold-change}</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The data importation requires a fc.info file, containing
column indices of each information in the data file and some
additional information, in the following order:
[0] Condition [1] Filename [2] Locus_tag column
[3] FC column [4] Separator [5] Startline [6] P-value column</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic annotation. If no annotation is
loaded, the load_annotation method with the default “sequence.gff3”
file is computed. To use another annotation, please load an
annotation to your Transcriptome instance with the following commands
before using this method:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">Genome</span><span class="o">,</span> <span class="nn">Transcriptome</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span> <span class="o">=</span> <span class="n">Transcriptome</span><span class="o">.</span><span class="n">Transcriptome</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="n">tr</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_annotation</span><span class="p">(</span><span class="n">annot_file</span><span class="o">=</span><span class="n">chosen_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">genes</span>
</pre></div>
</div>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span> <span class="o">=</span> <span class="n">Transcriptome</span><span class="o">.</span><span class="n">Transcriptome</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">load_fc_pval</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="s2">&quot;b0002&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">fc_pval</span>
<span class="go">{&#39;WT&#39;: (-1.786, 0.003125),</span>
<span class="go"> &#39;osmotic&#39;: (-1.81603942323042, 0.0),</span>
<span class="go"> &#39;heat&#39;: (1.62203513077067, 0.0)}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Transcriptome.Transcriptome.compute_state_from_fc">
<span class="sig-name descname"><span class="pre">compute_state_from_fc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thresh_pval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thresh_fc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Transcriptome.Transcriptome.compute_state_from_fc" title="Permalink to this definition"></a></dt>
<dd><p>Loads Fold-changes and p-values data and computes genes state from
these data. If no p-values is available in the data file, the default
value of 0 is assigned to each gene’s pvalue.</p>
<dl class="simple">
<dt>A gene is considered:</dt><dd><ul class="simple">
<li><p><cite>activated</cite> if its FC is above the FC threshold given as argument
and its pvalue is below the pvalue threshold given as
argument ie. FC &gt; thresh_FC and pval &lt; thresh_pval</p></li>
<li><p><cite>repressed</cite> if its FC is below the opposite of the FC threshold
and its pvalue is below the pvalue threshold
ie. FC &lt; -thresh_FC and pval &lt; thresh_pval</p></li>
<li><p><cite>not affected</cite> either if its pvalue is above the threshold,
or if its FC is between the - thresh_FC and + thresh_FC</p></li>
</ul>
</dd>
</dl>
<p>Creates:</p>
<blockquote>
<div><ul>
<li><dl class="simple">
<dt>self.statesFC (dict. of dict.)</dt><dd><p>New attribute of the  Transcriptome instance.
Dictionary containing one subdictionary per condition
listed in fc.info. Each subdictionary contains the list of
genes corresponding to  each state (‘act’, ‘rep’, ‘non’ or
‘null’).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.genes[locus].fc_pval (dict.)</dt><dd><p>new attribute of Gene instances
related to the Transcriptome instance given as argument.
Dictionary of shape {condition: (FC,pvalue)}.</p>
</dd>
</dl>
</li>
<li><dl>
<dt>self.genes[locus].state (dict.)</dt><dd><p>new attribute of Gene instances
related to the Transcriptome instance given as argument.
Dictionary of shape {condition: state} with state either</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>act</cite> if the gene is activated</p></li>
<li><p><cite>rep</cite> if the gene is repressed</p></li>
<li><p><cite>non</cite> if the gene is not affected</p></li>
<li><p><cite>null</cite> if the gene is not present in the data</p></li>
</ul>
</div></blockquote>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>thresh_pval</strong> (<em>Float</em>) – pvalue threshold used for the genes
classification</p></li>
<li><p><strong>thresh_fc</strong> (<em>Float</em>) – fold-change threshold used for the genes
classification</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The FC and pvalues data importation requires an FC.info file, in the
fold_changes directory, containing column indices of each information
in the data file, in the following order:
[0] Condition [1] Filename [2] Locus_tag column [3] FC columns
[4] Separator [5] File start line [6] P-value column
See load_fc_pval for more details about the data importation.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic annotation. If no annotation is
loaded, the load_annotation method with the default “sequence.gff3”
file is computed. To use another annotation, please load an
annotation to your Transcriptome instance with the following commands
before using this method:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">Transcriptome</span><span class="o">,</span> <span class="nn">Genome</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span> <span class="o">=</span> <span class="n">Transcriptome</span><span class="o">.</span><span class="n">Transcriptome</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="n">tr</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_annotation</span><span class="p">(</span><span class="n">annot_file</span><span class="o">=</span><span class="n">chosen_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">genes</span>
</pre></div>
</div>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span> <span class="o">=</span> <span class="n">Transcriptome</span><span class="o">.</span><span class="n">Transcriptome</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">compute_state_from_fc</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="s1">&#39;b0001&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">state</span>
<span class="go">{&#39;osmotic&#39;: &#39;rep&#39;, &#39;acidic_1mn&#39;: &#39;act&#39;}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="s1">&#39;b0001&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fc_pval</span>
<span class="go">{&#39;osmotic&#39;: (-1.73717046009437, 0.0), &#39;acidic_1mn&#39;: (1.73, 0.0)}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">statesFC</span><span class="p">[</span><span class="s1">&#39;osmotic&#39;</span><span class="p">][</span><span class="s1">&#39;rep&#39;</span><span class="p">]</span>
<span class="go">[&#39;b0001&#39;, &#39;b0002&#39;, &#39;b0003&#39;, &#39;b0004&#39;,...]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Transcriptome.Transcriptome.load_rnaseq_cov">
<span class="sig-name descname"><span class="pre">load_rnaseq_cov</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compute_from_bam</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Transcriptome.Transcriptome.load_rnaseq_cov" title="Permalink to this definition"></a></dt>
<dd><p>load_rnaseq_cov loads a RNASeq coverage to a Transcriptome instance
either from:</p>
<blockquote>
<div><ul class="simple">
<li><p>.npz files (described in a cov.info file and computed when new
data are loaded for the first time)  which are in the /rnaseq_cov/
directory,</p></li>
<li><p>coverage files which are in the /rnaseq_cov/ directory and are
described in cov_txt.info file,</p></li>
<li><p>paired-end .bam reads files which are in the /rnaseq_reads/
directory  and are treated with
useful_functions_transcriptome.cov_from_reads function.</p></li>
</ul>
</div></blockquote>
<dl class="simple">
<dt>Creates 2 new attributes of the Transcriptome instance:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.rnaseq_cov_pos:</dt><dd><p>dictionary of shape {condition: cov+}
with cov+ an array containing one signal data per genomic
position for the + strand (forward coverage)</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.rnaseq_cov_neg:</dt><dd><p>idem for the - strand (reverse coverage)</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cond</strong> (<em>Optional</em><em> [</em><em>list of str.</em><em>]</em>) – selection of one or several
conditions (1 condition corresponds to 1 data file).
By default: cond =’all’ ie all available coverages are
loaded.</p></li>
<li><p><strong>compute_from_bam</strong> (<em>Boolean.</em>) – if True, computes, using
useful_functions_transcriptome.cov_from_reads, coverage
from paired-end .bam reads files that are, with a
bam_files.info file, in the /rnaseq_reads/ directory.
WARNING: Works only with paired-end files</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To use directly new coverages data, both forward and reverse coverage
data files are required. These files contain the coverage for each
genomic position (one line = one genomic position, no position can be
ommited). The importation of these new data requires a cov_txt.info
file, in the /rnaseq_cov/ directory, containing the following
information:
[0] Condition
[1] Reverse coverage filename
[2] Forward coverage filename
[3] Startline (has to be the same for both coverage files)
[4] Column containing the coverage data</p>
<p>Coverages from the reverse file will be added to the Transcriptome
instance as “rnaseq_cov_neg” attribute and coverages from the forward
file will be the “rnaseq_cov_pos”attribute.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To compute coverages data from reads files, the data files and a
bam_files.info file have to be in the /rnaseq_reads/ directory.
The bam_files.info files contains:
[0] Condition [1] Reads filename</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>During the first importation of new coverages data, a .npy file
containing the data is created and will enable a faster data loading
for the next data importation. A cov.info file containing information
for this fast importation is also in the /rnaseq_cov/ directory and
contains in the following columns:
[0] Condition [1] Coverage filename
This cov.info file automatically completes itself when the data are
loaded for the first time.</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span> <span class="o">=</span> <span class="n">Transcriptome</span><span class="o">.</span><span class="n">Transcriptome</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="go"># To load only the coverage for one condition named &quot;WT&quot;</span>
<span class="go"># in the cov.info or cov_txt.info file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">load_rnaseq_cov</span><span class="p">([</span><span class="s2">&quot;WT&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">rnaseq_cov_neg</span><span class="p">[</span><span class="s2">&quot;WT&quot;</span><span class="p">]</span>
<span class="go">array([0., 0., 0., ..., 0., 0., 0.])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">rnaseq_cov_pos</span><span class="p">[</span><span class="s2">&quot;WT&quot;</span><span class="p">]</span>
<span class="go">array([0., 0., 0., ..., 0., 0., 0.])</span>
<span class="go"># To compute the coverage from a new .bam files obtained from</span>
<span class="go"># paired-end data for example with: tophat2 genome/MG1655</span>
<span class="go"># Test_1.fastq.gz Test_2.fastq.gz</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span> <span class="o">=</span> <span class="n">Transcriptome</span><span class="o">.</span><span class="n">Transcriptome</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">load_rnaseq_cov</span><span class="p">(</span><span class="n">compute_from_bam</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">rnaseq_cov_pos</span><span class="p">[</span><span class="s2">&quot;Test&quot;</span><span class="p">]</span>
<span class="go">array([10., 10., 10., ..., 0., 0., 0.])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">rnaseq_cov_pos</span><span class="p">[</span><span class="s2">&quot;Test&quot;</span><span class="p">]</span>
<span class="go">array([0., 0., 0., ..., 20., 20., 20.])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Transcriptome.Transcriptome.load_cov_start_end">
<span class="sig-name descname"><span class="pre">load_cov_start_end</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">compute_from_bam</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Transcriptome.Transcriptome.load_cov_start_end" title="Permalink to this definition"></a></dt>
<dd><p>Useful to locate TSS and TTS, load_cov_start_end loads the density of
RNA fragment starts and ends from .npz files described in
cov_start_stop.info, in the /rnaseq_cov/ directory.
.npz files and .info files are obtained using the
useful_functions_transcriptome.cov_start_stop_from_reads</p>
<dl class="simple">
<dt>Creates 2 new attributes of the Transcriptome instance:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.cov_start (Dict. of dict.)</dt><dd><p>dictionary containing one subdictionary per condition.
Each subdictionary has 2 keys:
“0” for the start positions on the - strand and
“1” for the start positions on the + strand
and the subdictionary values are arrays containing
all start positions on the corresponding strand</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>self.cov_end (Dict. of dict.)</dt><dd><p>idem for the end positions</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>compute_from_bam</strong> (<em>Boolean</em>) – if True, computes, using
useful_functions_transcriptome.cov_start_stop_from_reads,
coverage from paired-end .bam reads files that are, with
a bam_files.info file, in the /rnaseq_reads/ directory.</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To compute new coverages data from paired-end .bam reads files, the
data files and a bam_files.info file have to be in the /rnaseq_reads/
directory and the input argument “compute_from_bam” has to be set to
“True”.The bam_files.info files contains:
[0] Condition [1] Reads filename
During the importation of these new data, a corresponding .npz files
is created and the cov_start_stop.info file automatically completes
itself. For the next importations of these data, only the npz. and
the cov_start_stop.info will be loaded, thus allowing faster loading.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>compute_from_bam works only with paired-end files</p>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span> <span class="o">=</span> <span class="n">Transcriptome</span><span class="o">.</span><span class="n">Transcriptome</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">load_cov_start_end</span><span class="p">(</span><span class="n">compute_from_bam</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">cov_end</span><span class="p">[</span><span class="s2">&quot;WT_paired&quot;</span><span class="p">]</span>
<span class="go">{0: array([0, 0, 0, ..., 0, 0, 0]),</span>
<span class="go"> 1: array([0, 0, 0, ..., 0, 0, 0])}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Transcriptome.Transcriptome.compute_log2rpkm_from_cov">
<span class="sig-name descname"><span class="pre">compute_log2rpkm_from_cov</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">before</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Transcriptome.Transcriptome.compute_log2rpkm_from_cov" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Compute_log2rpkm_from_cov method:</dt><dd><ul class="simple">
<li><p>loads the RNASeq coverage (using load_rnaseq_cov method)</p></li>
<li><p>computes and loads RPKM value on the Gene instances
(using Gene.add_single_rpkm )</p></li>
<li><p>saves the log2(RPKM) data in a new .csv file
(1 file for all conditions, with 1 column per condition)</p></li>
<li><p>adds the new .csv file informations in the expression.info file
in the /expression/ directory</p></li>
</ul>
</dd>
<dt>Creates:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>self.genes[locus].rpkm (dict.):</dt><dd><p>new attribute of Gene instances related to the
Transcriptome instance given as argument.
Dictionary of shape {condition: RPKM value}, containing
the RPKM for each computed condition.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>log2rpkm_from_cov_{datetime.now()()}.csv: csv file containing the</dt><dd><p>log2(RPKM) values (1 row per gene, 1 column per condition).
If a RPKM is equal to 0, log2RPKM is set to 1.
This file is saved in the /expression/ directory and is
automatically added in the expression.info file.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cond</strong> (<em>list of str.</em>) – selection of one or several RNASeq coverage condition
By default: cond =’all’ ie all available coverages are
converted. All selected conditions have to be listed in
cov.info file. If one of the chosen condition is already
registered in the expression.info file in the expression/
directory, this condition is ignored.</p></li>
<li><p><strong>before</strong> (<em>int.</em>) – number of bps before the gene start to take into account</p></li>
</ul>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method needs a genomic annotation. If no annotation is
loaded, the load_annotation method with the default “sequence.gff3”
file is computed. To use another annotation, please load an
annotation to your Transcriptome instance with the following commands
before using this method:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">Genome</span><span class="o">,</span> <span class="nn">Transcriptome</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span> <span class="o">=</span> <span class="n">Transcriptome</span><span class="o">.</span><span class="n">Transcriptome</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">Genome</span><span class="p">(</span><span class="n">tr</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">load_annotation</span><span class="p">(</span><span class="n">annot_file</span><span class="o">=</span><span class="n">chosen_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="o">.</span><span class="n">tr</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">genes</span>
</pre></div>
</div>
</div>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span> <span class="o">=</span> <span class="n">Transcriptome</span><span class="o">.</span><span class="n">Transcriptome</span><span class="p">(</span><span class="s2">&quot;ecoli&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">compute_log2rpkm_from_cov</span><span class="p">(</span><span class="n">cond</span><span class="o">=</span><span class="s2">&quot;WT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tr</span><span class="o">.</span><span class="n">genes</span><span class="p">[</span><span class="s2">&quot;b0002&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rpkm</span><span class="p">[</span><span class="s2">&quot;WT&quot;</span><span class="p">]</span>
<span class="go">142</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</section>
<section id="module-useful_functions_transcriptome">
<span id="useful-functions"></span><h2>useful functions<a class="headerlink" href="#module-useful_functions_transcriptome" title="Permalink to this heading"></a></h2>
<p>Functions called by Transcriptome methods</p>
<dl class="py function">
<dt class="sig sig-object py" id="useful_functions_transcriptome.add_expression_to_genes">
<span class="sig-prename descclassname"><span class="pre">useful_functions_transcriptome.</span></span><span class="sig-name descname"><span class="pre">add_expression_to_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genes_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cond</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tag_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expr_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_log</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separator</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#useful_functions_transcriptome.add_expression_to_genes" title="Permalink to this definition"></a></dt>
<dd><p>Called by load_expression, adds expression data to Gene objects by
parsing a file with one column containing the locus tags of the genes and
one containing the expression data.</p>
<p>Creates a new attribute “expression” of the Gene objects contained in
the genes_dict given as argument. This attribute is a dictionary of shape
{condition: expression level [float.]}</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genes_dict</strong> (<em>dict.</em>) – dictionary of shape {locus: Gene object}</p></li>
<li><p><strong>cond</strong> (<em>str.</em>) – condition name</p></li>
<li><p><strong>filename</strong> (<em>str.</em>) – path to the file containing the expression data</p></li>
<li><p><strong>tag_col</strong> (<em>int.</em>) – index of the column containing the locus tags in the
file</p></li>
<li><p><strong>expr_col</strong> (<em>int.</em>) – index of the column containing the expression data</p></li>
<li><p><strong>is_log</strong> (<em>bool.</em>) – True if the data are already in log2, such as log2RPKM.
False if the data are RPKM.</p></li>
<li><p><strong>separartor</strong> (<em>str.</em>) – file separator</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Dict. of shape {‘locus’:Gene object} with in key the locus</dt><dd><p>tags associated to an expression data and in value the gene
object associated to each of these locus tags.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Column numbering starts at 0.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="useful_functions_transcriptome.load_fc_pval_cond">
<span class="sig-prename descclassname"><span class="pre">useful_functions_transcriptome.</span></span><span class="sig-name descname"><span class="pre">load_fc_pval_cond</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genes_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tag_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fc_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separator</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_line</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p_val_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#useful_functions_transcriptome.load_fc_pval_cond" title="Permalink to this definition"></a></dt>
<dd><p>Called by load_fc_pval, allows expression data to be loaded by specifying
a file containing the following information (one column for each type
of information): locus tags of the genes, log2(Fold-Changes) and p-value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genes_dict</strong> (<em>dict.</em>) – dictionary of shape {locus: Gene object}</p></li>
<li><p><strong>filename</strong> (<em>str.</em>) – path to the file containing the expression data</p></li>
<li><p><strong>cond</strong> (<em>str.</em>) – condition name</p></li>
<li><p><strong>tag_col</strong> (<em>int.</em>) – index of the column containing the locus tags in the
file</p></li>
<li><p><strong>fc_col</strong> (<em>int.</em>) – index of the column containing the log2FC</p></li>
<li><p><strong>separartor</strong> (<em>str.</em>) – file separator</p></li>
<li><p><strong>start_line</strong> (<em>int.</em>) – file start line</p></li>
<li><p><strong>p_val_col</strong> (<em>int.</em>) – index of the column containing the p-values in the
file. If no index is indicated (default: None), the p-values
will be assigned to 0 for all genes.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>locus tags of all genes having a valid log2FC</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>List</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Column numbering starts at 0.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="useful_functions_transcriptome.process_bam_paired_end">
<span class="sig-prename descclassname"><span class="pre">useful_functions_transcriptome.</span></span><span class="sig-name descname"><span class="pre">process_bam_paired_end</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tr</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#useful_functions_transcriptome.process_bam_paired_end" title="Permalink to this definition"></a></dt>
<dd><p>Called by load_cov_start_end and load_rnaseq_cov, process_bam_paired_end
converts paired-end .bam files in .npy files containing the fragments.
These files will enable a faster data loading for the next data
importation with load_cov_start_end or load_rnaseq_cov.</p>
<dl class="simple">
<dt>Creates:</dt><dd><ul class="simple">
<li><p>For each condition listed in the bam_files.info file, this function
creates a npz file containing the coverage data on both DNA strands
(Rpos.npy and Rneg.npy). These files are saved in the /rnaseq_reads/
directory</p></li>
<li><p>A reads.info file containing the information required to associate
each .npz  file to a condition name is also created (or completed if
this file already exists) in the /rnaseq_reads/ directory. This file
contains the following columns:
[0] Condition [1] Reads filename [2] BAM filename</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>tr</strong> – Transcriptome instance</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The paired-end .bam reads files are, with a bam_files.info file, in the
/rnaseq_reads/ directory. The bam_files.info file contains the following
information:  [0] Condition [1] Reads filename for this condition</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Only for paired-end reads !</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="useful_functions_transcriptome.cov_from_reads">
<span class="sig-prename descclassname"><span class="pre">useful_functions_transcriptome.</span></span><span class="sig-name descname"><span class="pre">cov_from_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tr</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#useful_functions_transcriptome.cov_from_reads" title="Permalink to this definition"></a></dt>
<dd><p>Called by load_rnaseq_cov, cov_from_reads computes the coverage from
paired-end reads previously converted to .npz format (containing one .npy
per strand: Rpos.npy and Rneg.npy) with the process_bam_paired_end
function.</p>
<dl class="simple">
<dt>Creates:</dt><dd><ul class="simple">
<li><p>For each condition listed in the reads.info file, this function
creates a npz file containing the coverage data on both DNA strands:
cov_pos.npy and cov_neg.npy. These files are saved in the
/rnaseq_cov/ directory and will enable a faster data loading for the
next data importation with load_rnaseq_cov.</p></li>
<li><p>A cov.info file containing the information required to associate
each .npz  file to a condition name is also created (or completed if
this file already exists) in the /rnaseq_cov/ directory. This file
contains the following columns:  [0] Condition [1] Coverage filename</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>tr</strong> – Transcriptome instance</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The .npz files have to be, with a reads.info file (also created
by the process_bam_paired_end function), in the /rnaseq_reads/ directory.
This reads.info file contains at least the following information:
[0] Condition [1] Reads filename</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="useful_functions_transcriptome.cov_start_stop_from_reads">
<span class="sig-prename descclassname"><span class="pre">useful_functions_transcriptome.</span></span><span class="sig-name descname"><span class="pre">cov_start_stop_from_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tr</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#useful_functions_transcriptome.cov_start_stop_from_reads" title="Permalink to this definition"></a></dt>
<dd><p>Called by load_cov_start_end, cov_start_stop_from_reads computes the
density of RNA fragment starts and ends from paired-end reads previously
converted to .npz format (containing one .npy per strand: Rpos.npy and
Rneg.npy) with the process_bam_paired_end function.</p>
<dl class="simple">
<dt>Creates:</dt><dd><ul class="simple">
<li><p>For each condition listed in the reads.info file, this function
creates a npz file containing the density of RNA fragment starts and
ends on both DNA strands: cov_start_pos.npy, cov_start_neg.npy,
cov_end_pos.npy and cov_end_neg.npy.
These files are saved in the /rnaseq_cov/ directory and will enable
a faster data loading for the next data importation with the
cov_start_stop_from_reads function.</p></li>
<li><p>A cov_start_stop.info file containing the information required to
associate each .npz  file to a condition name is also created (or
completed if this file already exists) in the /rnaseq_cov/ directory.
This file contains the following columns:
[0] Condition [1] Coverage filename</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>tr</strong> – Transcriptome instance</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The .npz files have to be, with a reads.info file (also created by
process_bam_paired_end function), in the /rnaseq_reads/ directory.
This reads.info file contains at least the following information:
[0] Condition [1] Reads filename</p>
</div>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Genome.html" class="btn btn-neutral float-left" title="Genome" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Chipseq.html" class="btn btn-neutral float-right" title="Chipseq" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Pineau, Forquet, Meyer.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>